### PR TITLE
scrub bootstrap chroot accordingly

### LINKED ISSUE #351: scrub command doesn't remove bootstrap data
### Short description of the problem

`mock -r <chroot-name> --scrub all` doesn't remove bootstrap data

### Output of `rpm -q mock`

    mock-1.4.19-1.fc30.noarch

### Steps to reproduce issue

1. `mock -r fedora-rawhide-x86_64 --bootstrap-chroot --shell`
2. `mock -r fedora-rawhide-x86_64 --scrub all`
3. `ls /var/lib/mock/ /var/cache/mock/ |grep rawhide`

### Any additional notes

    [jkadlcik@localhost mock]$ mock -r fedora-rawhide-x86_64 --scrub all
    ...
    [jkadlcik@localhost mock]$ ls /var/lib/mock/ /var/cache/mock/ |grep rawhide
    fedora-rawhide-x86_64-bootstrap
    fedora-rawhide-x86_64-bootstrap

### PR DESCRIPTION
**Pull request by FrostyX**
Fix #351

The `Commands.scrub` method requires `bootstrap_chroot` object initialized
in order to work with it. I.e. it doesn't scrub a bootstrap chroot unless
the `--bootstrap-chroot` parameter is set. Creating its own object inside
the method would duplicate a lot of existing code and doing it without the
object and its methods and properties is IMHO not a good idea.

Since the workaround for the #351 is

    mock -r <chroot> --bootstrap-chroot --scrub=all

we can just simply fix our parser to pretend like `--bootstrap-chroot`
was set, when using `--scrub`.

### CODE REVIEW

**Comment by praiskup** [comment]
`--scrub` can be called with
`all|chroot|cache|root-cache|c-cache|yum-cache|dnf-cache`.

Usually, when I don't specify --scrub=all, I don't care about bootstrap
chroot (it doesn't make sense to e.g. re-download its dnf-cache for
bootstrap chroot, or to re-install the chroot.  Bootstrap is just
mediator that user shouldn't usually know about, it shouldn't affect the
build results.

I'm not sure about this, but I'd slightly tend to just clean the bootstrap
chroot if user specifies `--scrub=all`.  Even #351 explicitly mentions 
`--scrub=all`, so perhaps that would be enough?

**Comment by FrostyX** [comment]
Good thinking. I've updated the PR.

**Comment by FrostyX** [comment]
Although if you look into the `Commands.scrub` method, it explicitly defines `boostrap_buildroot` deletion for all possible options.

**Comment by praiskup** [comment]
Ah, good point.  This is not only about explicit `--bootstrap-chroot` option, but also about config options in /etc/mock.  So shouldn't we rather add the check for `all` into a different place?

**Comment by FrostyX** [comment]
> but also about config options in /etc/mock

But you don't set _scrubbing_ in `/etc/mock`. You can set `config_opts['use_bootstrap_container']` there, but it shouldn't matter, right?

**Comment by praiskup** [comment]
Btw., similar change was dda9cd7bdb33a6a87b0788a3.

**Comment by FrostyX** [comment]
I moved it out of the parser. Do you like this version better?

**Comment by praiskup** [comment]
Well, there's still `mock -r <chroot> --scrub=<anything-else-than-all> --bootstrap-container`.  I think this shouldn't apply to bootstrap?  What do you think?

**Comment by FrostyX** [comment]
> Well, there's still mock -r <chroot> --scrub=<anything-else-than-all> --bootstrap-container. I think this shouldn't apply to bootstrap? What do you think?

So you basically want to set `config_opts['use_bootstrap_container']` to `False` when `--scrub=<anything-else-than-all>` even though you explicitly use `--bootstrap-container`?

**Comment by praiskup** [comment]
I think that with `scrub` on (any param), we should set `config_opts['use_bootstrap_container'] = True` (hack to get bootstrap initialized), but we should do more fine-grained cleanup in `scrub()` method depending on what the `--scrub=ARG` argument is.

We could have also `--scrub=bootstrap` which would only scrub the bootstrap, for example.

**Comment by FrostyX** [comment]
I've updated the PR and now it:

- it doesn't care whether you pass `--no-bootstrap-chroot` or set `config_opts['use_bootstrap_container'] = False`. When you use `--scrub` parameter with any value, it initializes the `bootstrap_buildroot` object.
- it adds a `--scrub=bootstrap` option to remove just bootstrap chroot
- it removes bootstrap chroot data only when `bootstrap` or `all` is passed. Otherwise, it leaves it

File: mock/py/mockbuild/backend.py [hunk 1]
```diff
@@ -105,10 +105,12 @@ def scrub(self, scrub_opts):
                 self.plugins.call_hooks('clean')
                 if self.bootstrap_buildroot is not None:
                     self.bootstrap_buildroot.plugins.call_hooks('clean')
+
                 for scrub in scrub_opts:
                     self.plugins.call_hooks('scrub', scrub)
                     if self.bootstrap_buildroot is not None:
                         self.bootstrap_buildroot.plugins.call_hooks('scrub', scrub)
+
```
Comment (FrostyX) [inline mock/py/mockbuild/backend.py]: This whitespace change was intentional. Previously it was squashed so much together, that it was had to read

File: mock/py/mockbuild/backend.py [hunk 2]
```diff
@@ -119,34 +121,23 @@ def scrub(self, scrub_opts):
...
                     elif scrub in ['yum-cache', 'dnf-cache']:
                         self.buildroot.root_log.info("scrubbing yum-cache and dnf-cache for %s", self.config_name)
                         util.rmtree(os.path.join(self.buildroot.cachedir, 'yum_cache'), selinux=self.buildroot.selinux)
                         util.rmtree(os.path.join(self.buildroot.cachedir, 'dnf_cache'), selinux=self.buildroot.selinux)
-                        if self.bootstrap_buildroot is not None:
-                            util.rmtree(os.path.join(self.bootstrap_buildroot.cachedir, 'yum_cache'),
-                                        selinux=self.bootstrap_buildroot.selinux)
-                            util.rmtree(os.path.join(self.bootstrap_buildroot.cachedir, 'dnf_cache'),
-                                        selinux=self.bootstrap_buildroot.selinux)
+                    elif scrub == 'bootstrap' and self.bootstrap_buildroot is not None:
```
Comment (praiskup) [inline mock/py/mockbuild/backend.py]: Here should be something in `self.buildroot.root_log.info('...')`.

File: mock/docs/mock.1 [hunk 3]
```diff
@@ -166,7 +166,7 @@ package and branch can be defined with \fB\-\-scm\-option\fP arguments,
 see site\-defaults.cfg for more information.
 .TP
 \fB\-\-scrub\fR=\fITYPE\fP
-Completely remove the specified chroot or cache dir or all of the chroot and cache.  \fITYPE\fR is one of all, chroot, cache, root\-cache, c\-cache, yum\-cache or dnf\-cache. In fact, dnf\-cache is just alias for yum\-cache, and both remove Dnf and Yum cache.
+Completely remove the specified chroot or cache dir or all of the chroot and cache.  \fITYPE\fR is one of all, chroot, bootstrap, cache, root\-cache, c\-cache, yum\-cache or dnf\-cache. In fact, dnf\-cache is just alias for yum\-cache, and both remove Dnf and Yum cache.
```
Comment (praiskup) [inline mock/docs/mock.1]: I'd put the `bootstrap` as the last one, but this is detail.

**Comment by praiskup** [comment]
Thanks, this looks nice!

**Comment by praiskup** [comment]
I'd somewhat prefer to have this as one commit, IMO at least the first two commits aren't fine without each other.

File: mock/py/mockbuild/backend.py (hunk 2 above)
Comment (FrostyX) [reply mock/py/mockbuild/backend.py]: I totally forgot on that. Fixed. Thank you!

**Comment by FrostyX** [comment]
> I'd somewhat prefer to have this as one commit, IMO at least the first two commits aren't fine without each other.

I've squashed those two together

**Comment by praiskup** [comment]
+1

File: mock/py/mockbuild/backend.py [hunk 4]
```diff
@@ -119,34 +121,24 @@ def scrub(self, scrub_opts):
...
                         util.rmtree(os.path.join(self.buildroot.cachedir, 'ccache'), selinux=self.buildroot.selinux)
-                        if self.bootstrap_buildroot is not None:
-                            util.rmtree(os.path.join(self.bootstrap_buildroot.cachedir, 'ccache'),
-                                        selinux=self.bootstrap_buildroot.selinux)
                     elif scrub == 'root-cache':
                         self.buildroot.root_log.info("scrubbing root-cache for %s", self.config_name)
                         util.rmtree(os.path.join(self.buildroot.cachedir, 'root_cache'), selinux=self.buildroot.selinux)
-                        if self.bootstrap_buildroot is not None:
-                            util.rmtree(os.path.join(self.bootstrap_buildroot.cachedir, 'root_cache'),
-                                        selinux=self.bootstrap_buildroot.selinux)
```
Comment (xsuchy) [inline mock/py/mockbuild/backend.py]: I do not get why these lines are being removed for root-cache, c-cache, cache, and yum-cache and dnf-cache?

Comment (praiskup) [reply mock/py/mockbuild/backend.py]: That's what we think is actually desired;  people doing `--scrub=root-cache` unlikely want to also clean `root-cache` of bootstrap chroot (they shouldn't mind, or even shouldn't basically know there's bootstrap chroot once it is default).

Comment (xsuchy) [reply mock/py/mockbuild/backend.py]: But if they know and they want to delete it, then there will be no way

Comment (praiskup) [reply mock/py/mockbuild/backend.py]: Right, they'd have to do `--scrub=bootstrap`, but yes ... most valid would be to have `--scrub=bootstrap-root-cache`.

Comment (xsuchy) [reply mock/py/mockbuild/backend.py]: IIRC, the `--scrub=bootstrap` will not clean bootstrap root cache

**Comment by praiskup** [comment]
On Tuesday, October 15, 2019 10:12:34 AM CEST Miroslav Suchý wrote:
> IIRC, the `--scrub=bootstrap` will not clean bootstrap root cache

Sure, but not _only_ that.  It will clean everything.  The question what
granularity we want WRT bootstrap.

**Comment by xsuchy** [comment]
@praiskup if I am reading the code correctly https://github.com/rpm-software-management/mock/blob/7c816e7ed7092d10e7be1b8e58adddde008c9e4f/mock/py/mockbuild/buildroot.py#L681 then `bootstrap_buildroot.delete()` will not delete e.g., `/var/cache/mock/epel-7-x86_64-bootstrap/root_cache`

**Comment by praiskup** [comment]
Yes, `chroot.delete()` is equal to `--clean` option, and it doesn't clean
cache dir itself.  That's why `--scrub=all` calls `chroot.delete()` **and**
`util.rmtree(self.buildroot.cachedir,...)`.  The same happens with the new
`--scrub=bootstrap` (either '--scrub=all' == full cleanup, or nothing).

I think we don't need more granularity for bootstrap cleanup, but you may
have different POV.

File: mock/py/mockbuild/backend.py [hunk 5]
```diff
@@ -119,34 +121,24 @@ def scrub(self, scrub_opts):
...
                         util.rmtree(os.path.join(self.buildroot.cachedir, 'dnf_cache'), selinux=self.buildroot.selinux)
-                        if self.bootstrap_buildroot is not None:
-                            util.rmtree(os.path.join(self.bootstrap_buildroot.cachedir, 'yum_cache'),
-                                        selinux=self.bootstrap_buildroot.selinux)
-                            util.rmtree(os.path.join(self.bootstrap_buildroot.cachedir, 'dnf_cache'),
-                                        selinux=self.bootstrap_buildroot.selinux)
+                    elif scrub == 'bootstrap' and self.bootstrap_buildroot is not None:
+                        self.buildroot.root_log.info("scrubbing bootstrap for %s", self.config_name)
+                        self.bootstrap_buildroot.delete()
+                        util.rmtree(self.bootstrap_buildroot.cachedir, selinux=self.bootstrap_buildroot.selinux)
```
Comment (praiskup) [inline mock/py/mockbuild/backend.py]: ... here is the bootstrap cache cleanup.

Comment (xsuchy) [reply mock/py/mockbuild/backend.py]: I was blind.

**Pull request merged**